README.rst
==========
scrapy-hcf
==========

.. image:: https://travis-ci.org/scrapy-plugins/scrapy-hcf.svg?branch=master
    :target: https://travis-ci.org/scrapy-plugins/scrapy-hcf

.. image:: https://codecov.io/gh/scrapy-plugins/scrapy-hcf/branch/master/graph/badge.svg
  :target: https://codecov.io/gh/scrapy-plugins/scrapy-hcf


This Scrapy spider middleware uses the HCF backend from Scrapinghub's
Scrapy Cloud service to retrieve the new urls to crawl
and store back the links extracted.


Installation
============

Install scrapy-hcf using ``pip``::

    $ pip install scrapy-hcf


Configuration
=============

To activate this middleware it needs to be added to the ``SPIDER_MIDDLEWARES``
dict, i.e::

    SPIDER_MIDDLEWARES = {
        'scrapy_hcf.HcfMiddleware': 543,
    }

And the following settings need to be defined:

``HS_AUTH``
    Scrapy Cloud API key

``HS_PROJECTID``
    Scrapy Cloud project ID (not needed if the spider is ran on dash)

``HS_FRONTIER``
    Frontier name.

``HS_CONSUME_FROM_SLOT``
    Slot from where the spider will read new URLs.

Note that ``HS_FRONTIER`` and ``HS_CONSUME_FROM_SLOT`` can be overriden
from inside a spider using the spider attributes ``hs_frontier``
and ``hs_consume_from_slot`` respectively.

The following optional Scrapy settings can be defined:

``HS_ENDPOINT``
    URL to the API endpoint, i.e: http://localhost:8003.
    The default value is provided by the python-hubstorage package.

``HS_MAX_LINKS``
    Number of links to be read from the HCF, the default is 1000.

``HS_START_JOB_ENABLED``
    Enable whether to start a new job when the spider finishes.
    The default is ``False``

``HS_START_JOB_ON_REASON``
    This is a list of closing reasons,
    if the spider ends with any of these reasons a new job will be started
    for the same slot. The default is ``['finished']``

``HS_NUMBER_OF_SLOTS``
    This is the number of slots that the middleware will use to store the new links.
    The default is 8.


Usage
=====

The following keys can be defined in a Scrapy Request meta in order to control the behavior
of the HCF middleware:

``'use_hcf'``
    If set to ``True`` the request will be stored in the HCF.

``'hcf_params'``
    Dictionary of parameters to be stored in the HCF with the request fingerprint

    ``'qdata'``
        data to be stored along with the fingerprint in the request queue

    ``'fdata'``
        data to be stored along with the fingerprint in the fingerprint set

    ``'p'``
        Priority - lower priority numbers are returned first. The default is 0

The value of ``'qdata'`` parameter could be retrieved later using
``response.meta['hcf_params']['qdata']``.

The spider can override the default slot assignation function by setting the
spider ``slot_callback`` method to a function with the following signature::

       def slot_callback(request):
           ...
           return slot


CHANGES.rst
Changes
=======


1.0.0 (2016-07-18)
------------------

Initial release.

This version is functionally equivalent to scrapylib's v1.7.0
``scrapylib.hcf.HcfMiddleware``.



tests/test_hcf.py
import os
import hashlib
import unittest

from scrapy.http import Request, Response
from scrapy.spider import Spider
from scrapy.utils.test import get_crawler
from scrapy.exceptions import NotConfigured
from hubstorage import HubstorageClient

from scrapy_hcf import HcfMiddleware

HS_ENDPOINT = os.getenv('HS_ENDPOINT', 'http://localhost:8003')
HS_AUTH = os.getenv('HS_AUTH')


@unittest.skipUnless(HS_AUTH, 'No valid hubstorage credentials set')
class HcfTestCase(unittest.TestCase):

    hcf_cls = HcfMiddleware

    projectid = '2222222'
    spidername = 'hs-test-spider'
    frontier = 'test'
    slot = '0'
    number_of_slots = 1

    @classmethod
    def setUpClass(cls):
        cls.endpoint = HS_ENDPOINT
        cls.auth = HS_AUTH
        cls.hsclient = HubstorageClient(auth=cls.auth, endpoint=cls.endpoint)
        cls.project = cls.hsclient.get_project(cls.projectid)
        cls.fclient = cls.project.frontier

    @classmethod
    def tearDownClass(cls):
        cls.project.frontier.close()
        cls.hsclient.close()

    def setUp(self):
        class TestSpider(Spider):
            name = self.spidername
            start_urls = [
                'http://www.example.com/'
            ]

        self.spider = TestSpider()
        self.hcf_settings = {'HS_ENDPOINT': self.endpoint,
                             'HS_AUTH': self.auth,
                             'HS_PROJECTID': self.projectid,
                             'HS_FRONTIER': self.frontier,
                             'HS_CONSUME_FROM_SLOT': self.slot,
                             'HS_NUMBER_OF_SLOTS': self.number_of_slots}
        self._delete_slot()

    def tearDown(self):
        self._delete_slot()

    def _delete_slot(self):
        self.fclient.delete_slot(self.frontier, self.slot)

    def _build_response(self, url, meta=None):
        return Response(url, request=Request(url="http://www.example.com/parent.html", meta=meta))

    def _get_crawler(self, settings=None):
        crawler = get_crawler(settings_dict=settings)
        # simulate crawler engine
        class Engine():
            def __init__(self):
                self.requests = []
            def schedule(self, request, spider):
                self.requests.append(request)
        crawler.engine = Engine()

        return crawler

    def test_not_loaded(self):
        crawler = self._get_crawler({})
        self.assertRaises(NotConfigured, self.hcf_cls.from_crawler, crawler)

    def test_start_requests(self):
        crawler = self._get_crawler(self.hcf_settings)
        hcf = self.hcf_cls.from_crawler(crawler)

        # first time should be empty
        start_urls = self.spider.start_urls
        new_urls = list(hcf.process_start_requests(start_urls, self.spider))
        self.assertEqual(new_urls, ['http://www.example.com/'])

        # now try to store some URLs in the hcf and retrieve them
        fps = [{'fp': 'http://www.example.com/index.html'},
               {'fp': 'http://www.example.com/index2.html'}]
        self.fclient.add(self.frontier, self.slot, fps)
        self.fclient.flush()
        new_urls = [r.url for r in hcf.process_start_requests(start_urls, self.spider)]
        expected_urls = [r['fp'] for r in fps]
        self.assertEqual(new_urls, expected_urls)
        self.assertEqual(len(hcf.batch_ids), 1)

    def test_spider_output(self):
        crawler = self._get_crawler(self.hcf_settings)
        hcf = self.hcf_cls.from_crawler(crawler)

        # process new GET request
        response = self._build_response("http://www.example.com/qxg1231")
        request = Request(url="http://www.example.com/product/?qxp=12&qxg=1231", meta={'use_hcf': True})
        outputs = list(hcf.process_spider_output(response, [request], self.spider))
        self.assertEqual(outputs, [])
        expected_links = {'0': set(['http://www.example.com/product/?qxp=12&qxg=1231'])}
        self.assertEqual(dict(hcf.new_links), expected_links)

        # process new POST request (don't add it to the hcf)
        response = self._build_response("http://www.example.com/qxg456")
        request = Request(url="http://www.example.com/product/?qxp=456", method='POST')
        outputs = list(hcf.process_spider_output(response, [request], self.spider))
        self.assertEqual(outputs, [request])
        expected_links = {'0': set(['http://www.example.com/product/?qxp=12&qxg=1231'])}
        self.assertEqual(dict(hcf.new_links), expected_links)

        # process new GET request (without the use_hcf meta key)
        response = self._build_response("http://www.example.com/qxg1231")
        request = Request(url="http://www.example.com/product/?qxp=789")
        outputs = list(hcf.process_spider_output(response, [request], self.spider))
        self.assertEqual(outputs, [request])
        expected_links = {'0': set(['http://www.example.com/product/?qxp=12&qxg=1231'])}
        self.assertEqual(dict(hcf.new_links), expected_links)

        # Simulate close spider
        hcf.close_spider(self.spider, 'finished')

    def test_close_spider(self):
        crawler = self._get_crawler(self.hcf_settings)
        hcf = self.hcf_cls.from_crawler(crawler)

        # Save 2 batches in the HCF
        fps = [{'fp': 'http://www.example.com/index_%s.html' % i} for i in range(0, 200)]
        self.fclient.add(self.frontier, self.slot, fps)
        self.fclient.flush()

        # Read the first batch
        start_urls = self.spider.start_urls
        new_urls = [r.url for r in hcf.process_start_requests(start_urls, self.spider)]
        expected_urls = [r['fp'] for r in fps]
        self.assertEqual(new_urls, expected_urls)

        # Simulate extracting some new urls
        response = self._build_response("http://www.example.com/parent.html")
        new_fps = ["http://www.example.com/child_%s.html" % i for i in range(0, 50)]
        for fp in new_fps:
            request = Request(url=fp, meta={'use_hcf': True})
            list(hcf.process_spider_output(response, [request], self.spider))
        self.assertEqual(len(hcf.new_links[self.slot]), 50)

        # Simulate emptying the scheduler
        crawler.engine.requests = []

        # Simulate close spider
        hcf.close_spider(self.spider, 'finished')
        self.assertEqual(len(hcf.new_links[self.slot]), 0)
        self.assertEqual(len(hcf.batch_ids), 0)

        # HCF must be have 1 new batch
        batches = [b for b in self.fclient.read(self.frontier, self.slot)]
        self.assertEqual(len(batches), 1)

    def test_hcf_params(self):
        crawler = self._get_crawler(self.hcf_settings)
        hcf = self.hcf_cls.from_crawler(crawler)

        # Simulate extracting some new urls and adding them to the HCF
        response = self._build_response("http://www.example.com/parent.html")
        new_fps = ["http://www.example.com/child_%s.html" % i for i in range(0, 5)]
        new_requests = []
        for fp in new_fps:
            hcf_params = {'qdata': {'a': '1', 'b': '2', 'c': '3'},
                          'fdata': {'x': '1', 'y': '2', 'z': '3'},
                          'p': 1}
            request = Request(url=fp, meta={'use_hcf': True, "hcf_params": hcf_params})
            new_requests.append(request)
            list(hcf.process_spider_output(response, [request], self.spider))
        expected = set(['http://www.example.com/child_4.html',
                        'http://www.example.com/child_1.html',
                        'http://www.example.com/child_0.html',
                        'http://www.example.com/child_3.html',
                        'http://www.example.com/child_2.html'])
        self.assertEqual(hcf.new_links[self.slot], expected)

        # Simulate close spider
        hcf.close_spider(self.spider, 'finished')

        # Similate running another spider
        start_urls = self.spider.start_urls
        stored_requests = list(hcf.process_start_requests(start_urls, self.spider))
        for a, b in zip(new_requests, stored_requests):
            self.assertEqual(a.url, b.url)
            self.assertEqual(a.meta.get('qdata'), b.meta.get('qdata'))

        # Simulate emptying the scheduler
        crawler.engine.requests = []

        # Simulate close spider
        hcf.close_spider(self.spider, 'finished')

    def test_spider_output_override_slot(self):
        crawler = self._get_crawler(self.hcf_settings)
        hcf = self.hcf_cls.from_crawler(crawler)

        def get_slot_callback(request):
            md5 = hashlib.md5()
            md5.update(request.url)
            digest = md5.hexdigest()
            return str(int(digest, 16) % 5)
        self.spider.slot_callback = get_slot_callback

        # process new GET request
        response = self._build_response("http://www.example.com/qxg1231")
        request = Request(url="http://www.example.com/product/?qxp=12&qxg=1231",
                          meta={'use_hcf': True})
        outputs = list(hcf.process_spider_output(response, [request], self.spider))
        self.assertEqual(outputs, [])
        expected_links = {'4': set(['http://www.example.com/product/?qxp=12&qxg=1231'])}
        self.assertEqual(dict(hcf.new_links), expected_links)

        # Simulate close spider
        hcf.close_spider(self.spider, 'finished')


scrapy_hcf/middleware.py
import os
import hashlib
import logging
from collections import defaultdict
from datetime import datetime
from scrapinghub import Connection
from scrapy import signals, log
from scrapy.exceptions import NotConfigured
from scrapy.http import Request
from hubstorage import HubstorageClient

DEFAULT_MAX_LINKS = 1000
DEFAULT_HS_NUMBER_OF_SLOTS = 8


class HcfMiddleware(object):

    def __init__(self, crawler):
        settings = crawler.settings
        self.hs_endpoint = settings.get("HS_ENDPOINT")
        self.hs_auth = self._get_config(settings, "HS_AUTH")
        self.hs_projectid = self._get_config(settings, "HS_PROJECTID", os.environ.get('SCRAPY_PROJECT_ID'))
        self.hs_frontier = self._get_config(settings, "HS_FRONTIER")
        self.hs_consume_from_slot = self._get_config(settings, "HS_CONSUME_FROM_SLOT")
        self.hs_number_of_slots = settings.getint("HS_NUMBER_OF_SLOTS", DEFAULT_HS_NUMBER_OF_SLOTS)
        self.hs_max_links = settings.getint("HS_MAX_LINKS", DEFAULT_MAX_LINKS)
        self.hs_start_job_enabled = settings.getbool("HS_START_JOB_ENABLED", False)
        self.hs_start_job_on_reason = settings.getlist("HS_START_JOB_ON_REASON", ['finished'])

        conn = Connection(self.hs_auth)
        self.panel_project = conn[self.hs_projectid]

        self.hsclient = HubstorageClient(auth=self.hs_auth, endpoint=self.hs_endpoint)
        self.project = self.hsclient.get_project(self.hs_projectid)
        self.fclient = self.project.frontier

        self.new_links = defaultdict(set)
        self.batch_ids = []

        crawler.signals.connect(self.close_spider, signals.spider_closed)

        # Make sure the logger for hubstorage.batchuploader is configured
        logging.basicConfig()

    def _get_config(self, settings, key, default=None):
        value = settings.get(key, default)
        if not value:
            raise NotConfigured('%s not found' % key)
        return value

    def _msg(self, msg, level=log.INFO):
        log.msg('(HCF) %s' % msg, level)

    def start_job(self, spider):
        self._msg("Starting new job for: %s" % spider.name)
        jobid = self.panel_project.schedule(
            spider.name,
            hs_consume_from_slot=self.hs_consume_from_slot,
            dummy=datetime.now()
        )
        self._msg("New job started: %s" % jobid)
        return jobid

    @classmethod
    def from_crawler(cls, crawler):
        return cls(crawler)

    def process_start_requests(self, start_requests, spider):

        self.hs_frontier = getattr(spider, 'hs_frontier', self.hs_frontier)
        self._msg('Using HS_FRONTIER=%s' % self.hs_frontier)

        self.hs_consume_from_slot = getattr(spider, 'hs_consume_from_slot', self.hs_consume_from_slot)
        self._msg('Using HS_CONSUME_FROM_SLOT=%s' % self.hs_consume_from_slot)

        self.has_new_requests = False
        for req in self._get_new_requests():
            self.has_new_requests = True
            yield req

        # if there are no links in the hcf, use the start_requests
        # unless this is not the first job.
        if not self.has_new_requests and not getattr(spider, 'dummy', None):
            self._msg('Using start_requests')
            for r in start_requests:
                yield r

    def process_spider_output(self, response, result, spider):
        slot_callback = getattr(spider, 'slot_callback', self._get_slot)
        for item in result:
            if isinstance(item, Request):
                request = item
                if request.meta.get('use_hcf', False):
                    if request.method == 'GET':  # XXX: Only GET support for now.
                        slot = slot_callback(request)
                        if not request.url in self.new_links[slot]:
                            hcf_params = request.meta.get('hcf_params')
                            fp = {'fp': request.url}
                            if hcf_params:
                                fp.update(hcf_params)
                            # Save the new links as soon as possible using
                            # the batch uploader
                            self.fclient.add(self.hs_frontier, slot, [fp])
                            self.new_links[slot].add(request.url)
                    else:
                        self._msg("'use_hcf' meta key is not supported for non GET requests (%s)" % request.url,
                                  log.ERROR)
                        yield request
                else:
                    yield request
            else:
                yield item

    def close_spider(self, spider, reason):
        # Only store the results if the spider finished normally, if it
        # didn't finished properly there is not way to know whether all the url batches
        # were processed and it is better not to delete them from the frontier
        # (so they will be picked by another process).
        if reason == 'finished':
            self._save_new_links_count()
            self._delete_processed_ids()

        # Close the frontier client in order to make sure that all the new links
        # are stored.
        self.fclient.close()
        self.hsclient.close()

        # If the reason is defined in the hs_start_job_on_reason list then start
        # a new job right after this spider is finished.
        if self.hs_start_job_enabled and reason in self.hs_start_job_on_reason:

            # Start the new job if this job had requests from the HCF or it
            # was the first job.
            if self.has_new_requests or not getattr(spider, 'dummy', None):
                self.start_job(spider)

    def _get_new_requests(self):
        """ Get a new batch of links from the HCF."""
        num_batches = 0
        num_links = 0
        for num_batches, batch in enumerate(self.fclient.read(self.hs_frontier, self.hs_consume_from_slot), 1):
            for fingerprint, data in batch['requests']:
                num_links += 1
                yield Request(url=fingerprint, meta={'hcf_params': {'qdata': data}})
            self.batch_ids.append(batch['id'])
            if num_links >= self.hs_max_links:
                break
        self._msg('Read %d new batches from slot(%s)' % (num_batches, self.hs_consume_from_slot))
        self._msg('Read %d new links from slot(%s)' % (num_links, self.hs_consume_from_slot))

    def _save_new_links_count(self):
        """ Save the new extracted links into the HCF."""
        for slot, new_links in self.new_links.items():
            self._msg('Stored %d new links in slot(%s)' % (len(new_links), slot))
        self.new_links = defaultdict(set)

    def _delete_processed_ids(self):
        """ Delete in the HCF the ids of the processed batches."""
        self.fclient.delete(self.hs_frontier, self.hs_consume_from_slot, self.batch_ids)
        self._msg('Deleted %d processed batches in slot(%s)' % (len(self.batch_ids),
                                                                self.hs_consume_from_slot))
        self.batch_ids = []

    def _get_slot(self, request):
        """ Determine to which slot should be saved the request."""
        md5 = hashlib.md5()
        md5.update(request.url)
        digest = md5.hexdigest()
        return str(int(digest, 16) % self.hs_number_of_slots)
